pageserver/src/tenant.rs

```diff
@@ -523,54 +503,38 @@ impl Tenant {
         .context("merge_local_remote_metadata")?
         .to_owned();
 
-        let timeline = {
+        let timeline = self.create_timeline_struct(
+            timeline_id,
+            up_to_date_metadata,
+            ancestor.clone(),
+            remote_client,
+            init_order,
+        )?;
+        let new_disk_consistent_lsn = timeline.get_disk_consistent_lsn();
+        anyhow::ensure!(
+            new_disk_consistent_lsn.is_valid(),
+            "Timeline {tenant_id}/{timeline_id} has invalid disk_consistent_lsn"
+        );
```

Comment:
This should be a post condition within the create_timeline_struct function

Reply:
It's not that easy.

There are two different types of users of `create_timeline_struct`.

The first type of user is for an already created timeline, e.g., during tenant load.

The second type of user is for timelines that are just being created: bootstrap, branch, import-from-basebackup.

* bootstrap & import-from-basebackup don't have a valid disk consistent lsn until we import the initdb dir or basebackup tar.
* for branch, we don't need to import anything, so, its disk_consistent lsn is valid

I'm planning more refactorings (see #4350 ) that will make it so that first and second type of users share the same loading path.

Until then, I think the PR is fine as it is.